### PR TITLE
Speed up ray verification and guard by verify_rays and add error testing

### LINKED ISSUE #16843: Simplify unique Ray verification and guard with parameter
## Reason
The verification of unique rays and ray IDs is a little sloppy. The ray ID verification does a global communication across all processors, which is unnecessary. In addition, it's not a bad idea to run it in optimized modes if ray verification is enabled.

## Design
Reduce communication for unique ray verification within the ray tracing module by removing the need for all-to-all communication. Also, guard what we can with the `verify_rays` parameter in RayTracingStudy.

## Impact
Faster Ray verification and more verification by default.

### PR DESCRIPTION
**Pull request by loganharbour**
Closes #16843 and #16848

### CODE REVIEW

File: modules/ray_tracing/test/src/userobjects/RepeatableRayStudyBaseTest.C [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+      _rays.push_back(ray);
+    }
+  }
+
+  if (parameters().isParamSetByUser("create_non_replicated_id_ray_pid"))
+  {
+    const auto pid = getParam<processor_id_type>("create_non_replicated_id_ray_pid");
+
+    mooseAssert(pid < n_processors(), "Invalid pid");
+    auto ray = acquireRegisteredRay("non_unique_ray");
```
Comment (GiudGiud) [inline modules/ray_tracing/test/src/userobjects/RepeatableRayStudyBaseTest.C:70]: "non_replicated_ray"

File: modules/ray_tracing/test/src/userobjects/RepeatableRayStudyBaseTest.C [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+    if (pid == _pid)
+    {
+      ray->setStart(_rays[0]->currentPoint());
+      ray->setStartingDirection(_rays[0]->direction());
+      _rays[0] = ray;
+    }
+  }
+
+  if (parameters().isParamSetByUser("create_non_replicated_ray_pid"))
```
Comment (GiudGiud) [inline modules/ray_tracing/test/src/userobjects/RepeatableRayStudyBaseTest.C:80]: check for param value

File: modules/ray_tracing/test/src/userobjects/RepeatableRayStudyBaseTest.C [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+    if (pid == _pid)
+    {
+      ray->setStart(_rays[0]->currentPoint());
+      ray->setStartingDirection(_rays[0]->direction());
+      _rays.push_back(ray);
+    }
+  }
+
+  if (parameters().isParamSetByUser("create_non_replicated_id_ray_pid"))
```
Comment (GiudGiud) [inline modules/ray_tracing/test/src/userobjects/RepeatableRayStudyBaseTest.C:65]: check for param value

File: modules/ray_tracing/test/src/userobjects/RepeatableRayStudyBaseTest.C (hunk 1 above)
Comment (GiudGiud) [reply modules/ray_tracing/test/src/userobjects/RepeatableRayStudyBaseTest.C:70]: or unique_ray

**Review by GiudGiud — CHANGES_REQUESTED** [review]

File: modules/ray_tracing/test/src/userobjects/RepeatableRayStudyBaseTest.C (hunk 2 above)
Comment (loganharbour) [reply modules/ray_tracing/test/src/userobjects/RepeatableRayStudyBaseTest.C:80]: Not needed. this is done below in an assertion (it's a test object)

File: modules/ray_tracing/test/src/userobjects/RepeatableRayStudyBaseTest.C (hunk 3 above)
Comment (loganharbour) [reply modules/ray_tracing/test/src/userobjects/RepeatableRayStudyBaseTest.C:65]: Not needed. this is done below in an assertion (it's a test object)

**Review by GiudGiud — APPROVED** [review]

**Pull request merged**